File: demos_code_J2_conception_objet/03_isp_roles/DemoIspAvant.java
public class DemoIspAvant {
    interface PaymentService {
        String find(String id);
        void capture(String id);
        void refund(String id);
        String exportCsv();
        void purgeAll();
    }

    static final class ReadOnlyPaymentView implements PaymentService {
        public String find(String id) { return "payment:" + id; }
        public void capture(String id) { throw new UnsupportedOperationException(); }
        public void refund(String id) { throw new UnsupportedOperationException(); }
        public String exportCsv() { throw new UnsupportedOperationException(); }
        public void purgeAll() { throw new UnsupportedOperationException(); }
    }

    public static void main(String[] args) {
        PaymentService view = new ReadOnlyPaymentView();
        System.out.println(view.find("P42"));
        try { view.refund("P42"); }
        catch (UnsupportedOperationException e) {
            System.out.println("Contrat mensonger : refund() n'est pas supporté");
        }
    }
}

File: Demo en Python/demos_code_J2_conception_objet_python/03_isp_roles/demo_isp_avant.py
from abc import ABC, abstractmethod


class PaymentService(ABC):
    @abstractmethod
    def find(self, payment_id: str) -> str: ...

    @abstractmethod
    def capture(self, payment_id: str) -> None: ...

    @abstractmethod
    def refund(self, payment_id: str) -> None: ...

    @abstractmethod
    def export_csv(self) -> str: ...

    @abstractmethod
    def purge_all(self) -> None: ...


class ReadOnlyPaymentView(PaymentService):
    def find(self, payment_id: str) -> str:
        return f"payment:{payment_id}"

    def capture(self, payment_id: str) -> None:
        raise NotImplementedError

    def refund(self, payment_id: str) -> None:
        raise NotImplementedError

    def export_csv(self) -> str:
        raise NotImplementedError

    def purge_all(self) -> None:
        raise NotImplementedError


if __name__ == "__main__":
    view: PaymentService = ReadOnlyPaymentView()
    print(view.find("P42"))
    try:
        view.refund("P42")
    except NotImplementedError:
        print("Contrat mensonger : refund() n'est pas supporté")
